src/validation.cpp

```diff
@@ -3920,13 +3922,13 @@ bool Chainstate::AcceptBlock(const std::shared_ptr<const CBlock>& pblock, BlockV
     // process an unrequested block if it's new and has enough work to
     // advance our tip, and isn't too many blocks ahead.
     bool fAlreadyHave = pindex->nStatus & BLOCK_HAVE_DATA;
-    bool fHasMoreOrSameWork = (m_chain.Tip() ? pindex->nChainWork >= m_chain.Tip()->nChainWork : true);
+    bool fHasMoreOrSameWork = (ActiveTip() ? pindex->nChainWork >= ActiveTip()->nChainWork : true);
```

Comment:
The idea here is that this check is part of a heuristic that we use when we receive a block that was not requested.  For anti-DoS reasons, back in #5875 we introduced logic to generally not process unrequested blocks, but we left in an optimization that was intended to not throw away blocks that would immediately advance our tip, in case we had some peer (like Matt's old fast-relay-network, if I remember right) that was sending such blocks (see eg comment here https://github.com/bitcoin/bitcoin/pull/5875#issuecomment-86743016).

My view is that we don't need to process unrequested blocks in general, and that if we do so it's just an optimization to make block relay at the tip work better.  We could probably add special-case logic to allow processing of unrequested blocks that we haven't yet validated that are ancestors of the ancestor block, but this doesn't strike me as something that is at all necessary (and perhaps not desirable?).